src/Symfony/Component/JsonPath/JsonCrawler.php

```diff
@@ -562,16 +824,228 @@ private function evaluateRecursive(mixed $value): array
+                        return true;
+                    }
+                }
+
+                if (TokenType::Name === $token->type && '*' === $token->value || TokenType::Recursive === $token->type) {
+                    return true;
+                }
+            }
+
+            return false;
+        } catch (InvalidJsonPathException) {
+            return true;
+        }
+    }
+
+    private function isNonSingularRelativeQuery(string $expr): bool
+    {
+        return preg_match('/@\[.*,.*]/', $expr) || '@.*' === $expr || preg_match('/@\[.*:.*]/', $expr);
+    }
+
+    private function findOperatorPosition(string $expr, string $op): int|false
+    {
+        $bracketDepth = 0;
+        $parenthesisDepth = 0;
+        $length = \strlen($expr);
+        $opLength = \strlen($op);
+
+        for ($i = 0; $i <= $length - $opLength; ++$i) {
+            $char = $expr[$i];
+
+            if ('[' === $char) {
+                ++$bracketDepth;
+            } elseif (']' === $char) {
+                --$bracketDepth;
+            } elseif ('(' === $char) {
+                ++$parenthesisDepth;
+            } elseif (')' === $char) {
+                --$parenthesisDepth;
+            } elseif (!$bracketDepth && !$parenthesisDepth && substr($expr, $i, $opLength) === $op) {
+                return $i;
+            }
+        }
+
+        return false;
+    }
+
+    private function validateFilterExpression(string $expr): void
+    {
+        if ($logicalOp = $this->findRightmostLogicalOperator($expr)) {
+            $this->validateFilterExpression(trim(substr($expr, 0, $logicalOp['position']))); // left
+            $this->validateFilterExpression(trim(substr($expr, $logicalOp['position'] + \strlen($logicalOp['operator'])))); // right
+
+            return;
+        }
+
+        foreach (self::COMPARISON_OPERATORS as $op => $len) {
+            if (str_contains($expr, $op)) {
+                if (false === $opPos = $this->findOperatorPosition($expr, $op)) {
+                    continue;
+                }
+
+                $left = trim(substr($expr, 0, $opPos));
+                $right = trim(substr($expr, $opPos + $len));
+
+                if (
+                    str_starts_with($left, '$') && $this->isNonSingularQuery($left)
+                    || str_starts_with($right, '$') && $this->isNonSingularQuery($right)
+                ) {
+                    throw new JsonCrawlerException($left, 'non-singular query is not comparable');
+                }
+
+                if (
+                    str_starts_with($left, '@') && $this->isNonSingularRelativeQuery($left)
+                    || str_starts_with($right, '@') && $this->isNonSingularRelativeQuery($right)
+                ) {
+                    throw new JsonCrawlerException($left, 'non-singular query is not comparable');
+                }
+
+                return;
+            }
+        }
+    }
+
     /**
      * Transforms JSONPath regex patterns to comply with RFC 9535.
```

Comment:
see my comment at https://github.com/symfony/symfony/pull/61517/files#r2324361812
+ the one below at https://github.com/symfony/symfony/pull/61517/files#r2324465838